### PR TITLE
Asnide 149 follow symbol is not working

### PR DESCRIPTION
**Pull request by psockoN7**
Data::TypeReference was removed from Data::TypeAssignment, and instead collection of user defined type references are stored in Data::File objects.

### CODE REVIEW

File: astxmlparser.cpp [hunk 1]
```diff
@@ -216,49 +214,81 @@ Data::SourceLocation AstXmlParser::readLocationFromAttributes()
...
+        { "IntegerType",       Data::Type::Kind::Integer },
+        { "RealType",          Data::Type::Kind::Real },
+        { "BitStringType",     Data::Type::Kind::BitString },
+        { "OctetStringType",   Data::Type::Kind::OctetString },
+        { "IA5StringType",     Data::Type::Kind::IA5String },
+        { "NumericStringType", Data::Type::Kind::NumericString },
+        { "EnumeratedType",    Data::Type::Kind::Enumerated },
+        { "ChoiceType",        Data::Type::Kind::Choice },
+        { "SequenceType",      Data::Type::Kind::Sequence },
+        { "SequenceOfType",    Data::Type::Kind::SequenceOf },
```
Comment (hcorg) [inline astxmlparser.cpp]: damn it's ugly :D "Type::Kind" - both words are quite similar and it makes it a little unreadable. Maybe Type should be a base class? Inheritance is better than enums..

File: data/builtintype.h [hunk 2]
```diff
@@ -24,16 +24,27 @@
...
 namespace Asn1Acn {
 namespace Internal {
-namespace Icons {
+namespace Data {
 
-QIcon iconForType(Data::Type type);
+class BuiltinType : public Type
+{
+public:
+    BuiltinType(const QString &name);
```
Comment (hcorg) [inline data/builtintype.h]: this string really hides an enum :(

Comment (psockoN7) [reply data/builtintype.h]: This is true, so do you think that it is better idea to add another inheritance level? This way Data::BuiltinType would become abstract, and classes which would represent different types  would be as follows (example for integer): 

```
class Integer : public BuiltinType
{
    public:
        Integer() = default;
        ~Integer() = default;

        QString label() const override { return QLatin1String("Integer"); }
        QIcon icon() const override { return createIcon("/path/to/icon/of/integer"); }
};

```

than in place where types are created (AstXmlParser) would be factory function 

```
std::unique_ptr<Data::Type> AstXmlParser::createType(const QString &type)
{
    if (type == QLatin1String("IntegerType"))
        return std::make_unique<Data::Type::Integer>();
    else if (type == QLatin1String("NullType"))
         return std::make_unique<Data::Type::Null>();
    .
    .
    .
    return nullptr;
}
```

?

Comment (hcorg) [reply data/builtintype.h]: hmmm that would be the most OOP solution, but generates a lot of classes (still - we might keep them in single BuiltinTypes.cpp/h) Wouldn't some enum in BuiltIn type itself be sufficient?

yet, if you have energy for that - go with full OOP ;)

Comment (psockoN7) [reply data/builtintype.h]: Ok, so I will go with OOP solution. I believe that there will be some pay off from this approach in future :)

**Comment by hcorg** [comment]
resolve conflicts plz

**Comment by psockoN7** [comment]
Merged master.

**Pull request merged**